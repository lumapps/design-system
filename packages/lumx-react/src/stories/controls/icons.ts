import {
    mdiAbTesting,
    mdiAbjadArabic,
    mdiAccount,
    mdiAccountBox,
    mdiAlert,
    mdiAlertCircle,
    mdiArrowDown,
    mdiArrowUp,
    mdiAtom,
    mdiBee,
    mdiBell,
    mdiBullhornOutline,
    mdiCheck,
    mdiCheckCircle,
    mdiChevronDown,
    mdiChevronLeft,
    mdiChevronRight,
    mdiChevronUp,
    mdiClose,
    mdiCloseCircle,
    mdiDelete,
    mdiDotsHorizontal,
    mdiDragVertical,
    mdiEarth,
    mdiEmail,
    mdiEye,
    mdiFileEdit,
    mdiFlag,
    mdiFolder,
    mdiFolderGoogleDrive,
    mdiFoodApple,
    mdiGoogleCirclesExtended,
    mdiHeart,
    mdiHome,
    mdiImageBroken,
    mdiInformation,
    mdiLink,
    mdiMagnifyMinusOutline,
    mdiMagnifyPlusOutline,
    mdiMenuDown,
    mdiMessageTextOutline,
    mdiMinus,
    mdiOpenInNew,
    mdiPauseCircleOutline,
    mdiPencil,
    mdiPlay,
    mdiPlayCircleOutline,
    mdiPlus,
    mdiRadioboxBlank,
    mdiRadioboxMarked,
    mdiReply,
    mdiSend,
    mdiStar,
    mdiTextBox,
    mdiTextBoxPlus,
    mdiTram,
    mdiTranslate,
    mdiViewList,
} from '@lumx/icons';

// Small selection of mdi icons to
export const iconArgType = {
    control: { type: 'select' },
    options: {
        mdiAbTesting,
        mdiAbjadArabic,
        mdiAccount,
        mdiAccountBox,
        mdiAlert,
        mdiAlertCircle,
        mdiArrowDown,
        mdiArrowUp,
        mdiAtom,
        mdiBee,
        mdiBell,
        mdiBullhornOutline,
        mdiCheck,
        mdiCheckCircle,
        mdiChevronDown,
        mdiChevronLeft,
        mdiChevronRight,
        mdiChevronUp,
        mdiClose,
        mdiCloseCircle,
        mdiDelete,
        mdiDotsHorizontal,
        mdiDragVertical,
        mdiEarth,
        mdiEmail,
        mdiEye,
        mdiFileEdit,
        mdiFlag,
        mdiFolder,
        mdiFolderGoogleDrive,
        mdiFoodApple,
        mdiGoogleCirclesExtended,
        mdiHeart,
        mdiHome,
        mdiImageBroken,
        mdiInformation,
        mdiLink,
        mdiMagnifyMinusOutline,
        mdiMagnifyPlusOutline,
        mdiMenuDown,
        mdiMessageTextOutline,
        mdiMinus,
        mdiOpenInNew,
        mdiPauseCircleOutline,
        mdiPencil,
        mdiPlay,
        mdiPlayCircleOutline,
        mdiPlus,
        mdiRadioboxBlank,
        mdiRadioboxMarked,
        mdiReply,
        mdiSend,
        mdiStar,
        mdiTextBox,
        mdiTextBoxPlus,
        mdiTram,
        mdiTranslate,
        mdiViewList,
    },
};
