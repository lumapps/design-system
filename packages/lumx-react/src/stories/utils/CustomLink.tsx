/* eslint-disable jsx-a11y/anchor-has-content,jsx-a11y/anchor-is-valid */
import React from 'react';

/**
 * Example custom link that can be used in linkAs props.
 */
export const CustomLink: any = ({ ...props }) => <a href="#" data-custom-link {...props} />;
