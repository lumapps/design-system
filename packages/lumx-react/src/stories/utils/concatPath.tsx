export type PathPart = string | number;
export type Path = PathPart | Array<PathPart>;

/**
 * Concat flatten object path
 *
 * @example concatPath('foo', 'bar') // => 'foo.bar'
 * @example concatPath(['foo', 0]) // => 'foo[0]'
 * @example concatPath('foo', 0, ['bar']) // => 'foo[0].bar'
 */
export const concatPath = (...prefix: Path[]) => {
    const [first, ...rest] = prefix.flat();
    return rest.reduce<string>((acc, part) => {
        if (typeof part === 'number') return `${acc}[${part}]`;
        return `${acc}.${part}`;
    }, String(first));
};
