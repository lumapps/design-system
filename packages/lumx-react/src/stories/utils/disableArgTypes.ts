export function disableArgTypes(argTypes: string[]) {
    return Object.fromEntries(argTypes.map((argType) => [argType, { table: { disable: true } }]));
}
