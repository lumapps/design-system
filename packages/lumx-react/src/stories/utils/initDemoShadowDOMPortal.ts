/** Init a demo shadow DOM to use as portal container */
export const initDemoShadowDOMPortal = () => {
    const div = document.createElement('div');
    document.body.appendChild(div);
    const container = div.attachShadow({ mode: 'open' });
    const style = document.createElement('style');
    style.innerText = `button { color: red; }`;
    container.appendChild(style);
    return { container, teardown: () => div.remove() };
};
