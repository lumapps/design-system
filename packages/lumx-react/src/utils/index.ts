/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
