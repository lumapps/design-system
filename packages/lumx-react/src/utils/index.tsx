export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
