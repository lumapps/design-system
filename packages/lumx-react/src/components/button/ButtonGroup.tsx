import React, { forwardRef } from 'react';

import classNames from 'classnames';

import { Comp, GenericProps, getRootClassName } from '@lumx/react/utils';

/**
 * Defines the props of the component
 */
export type ButtonGroupProps = GenericProps;

/**
 * Component display name.
 */
const COMPONENT_NAME = 'ButtonGroup';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<ButtonGroupProps> = {};

/**
 * ButtonGroup component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const ButtonGroup: Comp<ButtonGroupProps, HTMLDivElement> = forwardRef((props, ref) => {
    const { children, className, ...forwardedProps } = props;

    return (
        <div ref={ref} {...forwardedProps} className={classNames(className, CLASSNAME)}>
            {children}
        </div>
    );
});
ButtonGroup.displayName = COMPONENT_NAME;
ButtonGroup.className = CLASSNAME;
ButtonGroup.defaultProps = DEFAULT_PROPS;
