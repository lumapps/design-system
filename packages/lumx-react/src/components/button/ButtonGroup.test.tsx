import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';

import { mdiPlus } from '@lumx/icons';
import { Button, IconButton } from '@lumx/react';
import { commonTestsSuite, Wrapper } from '@lumx/react/testing/utils';
import { ButtonGroup, ButtonGroupProps } from './ButtonGroup';

const CLASSNAME = ButtonGroup.className as string;

type SetupProps = Partial<ButtonGroupProps>;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = ({ ...propsOverride }: SetupProps = {}, shallowRendering = true) => {
    const props: ButtonGroupProps = {
        children: (
            <>
                <Button>Label</Button>
                <IconButton label="Add" icon={mdiPlus} />
            </>
        ),
        ...propsOverride,
    };
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;
    const wrapper: Wrapper = renderer(<ButtonGroup {...props} />);

    return {
        group: wrapper.find(`.${CLASSNAME}`),
        props,
        wrapper,
    };
};

describe(`<${ButtonGroup.displayName}>`, () => {
    // 1. Test render via snapshot (default state of component).
    describe('Snapshots and structure', () => {
        it('should render correctly a group button', () => {
            const { group, wrapper } = setup();
            expect(wrapper).toMatchSnapshot();

            expect(group).toExist();
            expect(group).toHaveClassName(CLASSNAME);

            expect(group.children().length).toEqual(2);
        });
    });

    // 2. Test defaultProps value and important props custom values.
    // N/A

    // 3. Test events.
    // N/A

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    // N/A

    // 5. Test state.
    // N/A

    // Common tests suite.
    commonTestsSuite(setup, { className: 'group', prop: 'group' }, { className: CLASSNAME });
});
