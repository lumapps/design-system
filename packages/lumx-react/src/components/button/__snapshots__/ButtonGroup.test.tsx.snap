// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonGroup> Snapshots and structure should render correctly a group button 1`] = `
<div
  className="lumx-button-group"
>
  <Button
    emphasis="high"
    size="m"
    theme="light"
  >
    Label
  </Button>
  <IconButton
    emphasis="high"
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    label="Add"
    size="m"
    theme="light"
  />
</div>
`;
